### DISCUSSION TITLE
How does claude choose the correct service when there are multiple MCP servers

### DISCUSSION THREAD

**Discussion by just521lili** [modelcontextprotocol/modelcontextprotocol#1198] [Q&A - General]
### Pre-submission Checklist

- [X] I have checked that this question would not be more appropriate as an issue in a specific repository
- [X] I have searched existing discussions and documentation for answers

### Question Category

- [X] Protocol Specification
- [X] SDK Usage
- [X] Server Implementation
- [X] General Implementation
- [X] Documentation
- [X] Other

### Your Question

When I configure multiple MCP servers, how does claude determine which service to call when I input a sentence, does claude have fields such as desc as prompts, and how to configure it? Thank you

**Comment by badkk** [comment]
I guess you'll need to set up keyword triggers in your MCP server code. 

Like the `description` shows here: https://github.com/anjor/coinmarket-mcp-server/blob/568310e4aab30969125eaeae8f4834c43c10a1c8/src/coinmarket_service/server.py#L112

Correct me if i'm wrong @jspahrsummers

**Comment by Sudhanshugtm** [comment]
I generally mention in my prompt use xyz tool for this purpose

**Comment by headinthebox** [comment]
I have a similar question. What if there are multiple servers that expose tools with the same name. The console(https://console.anthropic.com/) does not allow you to add two tools with the same name (makes sense), but MCP must allow it.
Does it use the server name in the config file as the namespace for tool names?

**Reply by dsp-ant** [reply]
The behaviour for this is currently unspecified. Clients should likely track the tools that are added and if a conflict arises do one of the following:
* Disable the offending servers and inform the user.
* Rename the tools from one server by prefixing them with a unique server-id and informing the model about this. Notify the user.

**Comment by tadasant** [comment]
There was a similar question in the Anthropic Discord: https://discord.com/channels/1072196207201501266/1323580910079185008/1323580910079185008

Someone gave this answer:

![CleanShot 2025-01-05 at 17 09 40](https://github.com/user-attachments/assets/7eeda54c-b8bd-407a-9543-0e4ad582e8dd)

I'm not sure exactly where he pulled that level of detail, but it's similar to how Claude's [Tool Use](https://docs.anthropic.com/en/docs/build-with-claude/tool-use#tool-use-system-prompt) is documented:

> # Tool use system prompt
> When you call the Anthropic API with the tools parameter, we construct a special system prompt from the tool definitions, tool configuration, and any user-specified system prompt. The constructed prompt is designed to instruct the model to use the specified tool(s) and provide the necessary context for the tool to operate properly:

> In this environment you have access to a set of tools you can use to answer the user's question.
> {{ FORMATTING INSTRUCTIONS }}
> String and scalar parameters should be specified as is, while lists and objects should use JSON format. Note that spaces for string > values are not stripped. The output is not expected to be valid XML and is parsed with regular expressions.
> Here are the functions available in JSONSchema format:
> {{ TOOL DEFINITIONS IN JSON SCHEMA }}
> {{ USER SYSTEM PROMPT }}
> {{ TOOL CONFIGURATION }}

---

So I'd say the answer to your question

> how does claude determine which service to call when I input a sentence

is basically: the tools get stuffed into the system prompt, and from there the "selection" gets made the same way any prompt is processed; by the black box that is the foundational LLM.

I don't think Claude is designed to elegantly handle name clashes, and unless they change the way tool calling is implemented to be more structured in some way, that's a problem that is more likely to be solved at the MCP specification level than in Claude itself.

**Reply by ThoughtPhotography** [reply]
> I don't think Claude is designed to elegantly handle name clashes, and unless they change the way tool calling is implemented to be more structured in some way, that's a problem that is more likely to be solved at the MCP specification level than in Claude itself.

It doesn't handle name clashes well.
I have at least 3 servers installed with a **list_allowed_directories** tool.
But now instead of filesystem, which I've used 100% of the time, it might pick a different server.

In my opinion, Claude should prioritize based on the order servers appear in the .json, and/or include a drop-down for selecting the server as part of the permission request to resolve name conflicts. 

Also, it's probably a bad idea for developers to use identical tool names as official Anthropic servers.

And when using basic tool names like __read_file_,_ a suffix, or prefix should be included (e.g. _read_file_dc_ for desktop-commander)

This:
> "Use the list_allowed_directories_aidd tool to..."

Is much more convenient that this:
> "Use the list_allowed_directories tool from the aidd-ai-software-development-utilities server to..."

**Comment by zhjch05** [comment]
I think server name and description will work on client side. But when I try to make a proxy server that can aggregate other server tools, I cannot distinguish tool calls from client side (they just send in tool name, description and schema). I tried to hack by prefixing a namespace using server name but tool name has a regex validation of ^[a-zA-Z0-9_-]{1,64}$ so I cannot split out the server name as namespace.. Is there any support planned on explicit namespace of tools? Thanks.

**Reply by zhjch05** [reply]
sorry I think this is mentioned int he specification discussions https://github.com/modelcontextprotocol/specification/discussions/94

**Reply by headinthebox** [reply]
^[a-zA-Z0-9_-]{1,64}$

I ran into this as well, and I resorted to using maintaining a "symbol table" inside my proxy, and translating back and forth. It maps the pair of (server, function) names to a unique name that I give to Claude, and then when Claude tells me to call that function, I look up the pair of server and function to make the actual call.

**Comment by localden** [comment]
We've limited Discussions in this repository to meeting notes and are closing this thread. To continue the conversation, pick a channel from [Contributor Communication](https://modelcontextprotocol.io/community/communication). Thanks for posting.
